main.tf

```diff
@@ -0,0 +1,331 @@
+# DEPLOY A CLOUD SQL CLUSTER
+# This module deploys a Cloud SQL MySQL cluster. The cluster is managed by Google and automatically handles leader
+# election, replication, failover, backups, patching, and encryption.
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+terraform {
+  # This module is now only being tested with Terraform 1.0.x. However, to make upgrading easier, we are setting
+  # 0.12.26 as the minimum version, as that version added support for required_providers with source URLs, making it
+  # forwards compatible with 1.0.x code.
+  required_version = ">= 0.12.26"
+}
+
+# ------------------------------------------------------------------------------
+# PREPARE LOCALS
+#
+# NOTE: Due to limitations in terraform and heavy use of nested sub-blocks in the resource,
+# we have to construct some of the configuration values dynamically
+# ------------------------------------------------------------------------------
+
+locals {
+  # Determine the engine type
+  is_postgres = replace(var.engine, "POSTGRES", "") != var.engine
+  is_mysql    = replace(var.engine, "MYSQL", "") != var.engine
+
+  # Calculate actuals, so we get expected behavior for each engine
+  actual_binary_log_enabled     = local.is_postgres ? false : var.mysql_binary_log_enabled
+  actual_availability_type      = local.is_postgres && var.enable_failover_replica ? "REGIONAL" : "ZONAL"
+  actual_failover_replica_count = local.is_postgres ? 0 : var.enable_failover_replica ? 1 : 0
+}
+
+# ------------------------------------------------------------------------------
+# CREATE THE MASTER INSTANCE
+#
+# NOTE: We have multiple google_sql_database_instance resources, based on
+# HA and replication configuration options.
+# ------------------------------------------------------------------------------
+
+resource "google_sql_database_instance" "master" {
+  depends_on = [null_resource.dependency_getter]
+
+  provider         = google-beta
+  name             = var.name
+  project          = var.project
+  region           = var.region
+  database_version = var.engine
+
+  # Whether or not to allow Terraform to destroy the instance.
+  deletion_protection = var.deletion_protection
+
+  settings {
+    tier              = var.machine_type
+    activation_policy = var.activation_policy
+    disk_autoresize   = var.disk_autoresize
+
+    ip_configuration {
+      dynamic "authorized_networks" {
+        for_each = var.authorized_networks
+        content {
+          name  = lookup(authorized_networks.value, "name", null)
+          value = authorized_networks.value.value
+        }
+      }
+
+      ipv4_enabled    = var.enable_public_internet_access
+      private_network = var.private_network
+      require_ssl     = var.require_ssl
+    }
+
+    dynamic "location_preference" {
+      for_each = var.master_zone == null ? [] : [var.master_zone]
```

Comment:
use for_each = [var.master_zone] ans make default value of master_zone = **null** or **{}**